proposals/0021-root-signature-in-clang.md

```diff
@@ -0,0 +1,247 @@
+
+To ensure that a used resource has a binding in the root signature, this
+information should be accessible in the backend to make sure legal dxil is
+generated.
+
+A new AST node HLSLRootSignatureDecl will be added to represent the root
+signature in the AST.
+To avoid parsing root signature string twice, HLSLRootSignatureDecl will 
+save the parsed root siganture instead of the string.
+This will require more work for print and serialization the AST node.
+Since DirectX backend also need to print and serialzation root signature, 
+the code should be able to share between the frontend and backend.
+
+A HLSLRootSignatureAttr will be created when meet RootSignature attribute in
+HLSL.
+Because the RootSignature attribute in hlsl only have the string, it is easier
+to add a StringArgument to save the string first.
+A HLSLRootSignatureDecl will be created and added to the HLSLRootSignatureAttr 
+for diagnostic and saved to the HLSLEntryRootSignatureAttr for clang 
+code generation.
+The HLSLRootSignatureDecl will save StringLiteral instead StringRef for 
+diagnostic.
+
+In clang code generation, the HLSLRootSignatureAttr in AST will be translated
+into metadata to express the layout and things like static sampler, root 
+flags, space and NumDescriptors in LLVM IR.
+
+CGHLSLRuntime will generate metadata to link the metadata as root
+signature for given entry function.
+
+For case compile to a standalone root signature blob, the
+HLSLRootSignatureAttr will be bind to a fake empty entry.
+
+In the case of local or global root signatures, only the HLSLRootSignatureDecl 
+will be created.
+
+Additionally, a new AST node, SubobjectToExportsAssociationDecl, should be 
+added to handle the SubobjectToExportsAssociation statement in HLSL. 
+The SubobjectToExportsAssociationDecl will contain a HLSLRootSignatureDecl and 
+a list of FunctionDecls.
+
+The default root signature version number is 2 which maps to rootsig_1_1.
+With option -force-rootsig-ver rootsig_1_0 could make it to 1.
+
+## Detailed design
+
+
+* Example root signature
+
+```
+
+    #define RS "RootFlags( ALLOW_INPUT_ASSEMBLER_INPUT_LAYOUT | " \
+              "DENY_VERTEX_SHADER_ROOT_ACCESS), " \
+              "CBV(b0, space = 1, flags = DATA_STATIC), " \
+              "SRV(t0), " \
+              "UAV(u0), " \
+              "DescriptorTable( CBV(b1), " \
+              "                 SRV(t1, numDescriptors = 8, " \
+              "                     flags = DESCRIPTORS_VOLATILE), " \
+              "                 UAV(u1, numDescriptors = unbounded, " \
+              "                     flags = DESCRIPTORS_VOLATILE)), " \
+              "DescriptorTable(Sampler(s0, space=1, numDescriptors = 4)), " \
+              "RootConstants(num32BitConstants=3, b10), " \
+              "StaticSampler(s1)," \
+              "StaticSampler(s2, " \
+              "              addressU = TEXTURE_ADDRESS_CLAMP, " \
+              "              filter = FILTER_MIN_MAG_MIP_LINEAR )"
+
+```
```

Comment:
This isn't actually an example of the root signature - it's just a define that matches the syntax of a root signature. It'd be helpful for readers here to see the various ways that a root signature is defined. That is, we can specify the root signature as an attribute, via `-rootsig-define`, and using `GlobalRootSignature` or `LocalRootSignature` types (are there more?). It would be helpful to point these out, especially as a way to explain what exactly the HLSLRootSignature decl would mean.

Reply:
Added a Background part to show all these different root signatures.